Integrations/src/main/java/io/deephaven/integrations/python/PythonDeephavenSession.java

```diff
@@ -295,34 +274,49 @@ private Object maybeUnwrap(PyObject o) {
     }
 
     @Override
-    public Set<String> getVariableNames() {
-        return Collections.unmodifiableSet(scope.getKeys().collect(Collectors.toSet()));
+    protected Set<String> getVariableNames(Predicate<String> allowName) {
+        return PyLib.ensureGil(() -> scope.getKeys().filter(allowName).collect(Collectors.toUnmodifiableSet()));
     }
 
     @Override
-    public boolean hasVariableName(String name) {
+    protected boolean hasVariable(String name) {
         return scope.containsKey(name);
     }
 
     @Override
-    public synchronized void setVariable(String name, @Nullable Object newValue) {
-        final PyDictWrapper globals = scope.mainGlobals();
-        if (newValue == null) {
-            try {
-                globals.delItem(name);
-            } catch (KeyError key) {
-                // ignore
-            }
-        } else {
-            if (!(newValue instanceof PyObject)) {
-                newValue = PythonObjectWrapper.wrap(newValue);
+    protected synchronized Object setVariable(String name, @Nullable Object newValue) {
+        Object old = PyLib.ensureGil(() -> {
+            final PyDictWrapper globals = scope.mainGlobals();
+
+            if (newValue == null) {
+                try {
+                    return globals.unwrap().callMethod("pop", name);
+                } catch (KeyError key) {
+                    return null;
+                }
+            } else {
+                Object wrapped;
+                if (newValue instanceof PyObject) {
+                    wrapped = newValue;
+                } else {
+                    wrapped = PythonObjectWrapper.wrap(newValue);
+                }
+                // This isn't thread safe, we're relying on the GIL being kind to us (as we have historically done).
```

Comment:
Is this true, given that we've enclosing with `ensureGil`?

Reply:
Technically no, the GIL doesn't guarantee us practically anything even if this was one call instead of two - `__setitem__` is just a function that could be redefined in python in a way that lets it drop the gil while invoking it...

Practically, we hope that it continued to be kind to us.